langserver/workspace_refs.go

```diff
@@ -70,27 +72,36 @@ func (h *LangHandler) handleWorkspaceReferences(ctx context.Context, conn JSONRP
 				continue
 			}
 		}
+		unvendoredPackages[bpkg.ImportPath] = struct{}{}
 		pkgs = append(pkgs, pkg)
 	}
 	if len(pkgs) == 0 {
 		// occurs when the directory hint is present and matches no directories
 		// at all.
 		return []lspext.ReferenceInformation{}, nil
 	}
-	prog, err := h.workspaceRefsTypecheck(ctx, bctx, conn, fset, pkgs)
-	if err != nil {
-		return nil, err
-	}
 
-	// Collect dependency references.
-	results := refResultSorter{results: make([]lspext.ReferenceInformation, 0)}
-	par := parallel.NewRun(parallelism)
-	for _, pkg := range prog.Imported {
-		par.Acquire()
-		go func(pkg *loader.PackageInfo) {
-			defer par.Release()
+	// Collect dependency references in the AfterTypeCheck phase. This enables
+	// us to begin looking at packages as they are typechecked, instead of
+	// waiting for all packages to be typechecked (which is IO bound).
+	var (
+		results      = refResultSorter{results: make([]lspext.ReferenceInformation, 0)}
+		finishSignal = make(chan struct{})
+		started      uint32
+	)
+	afterTypeCheck := func(pkg *loader.PackageInfo, files []*ast.File) {
+		_, interested := unvendoredPackages[pkg.Pkg.Path()]
+		if !interested {
+			return
+		}
+
+		// Do not block the type-checker, but also avoid running with unbounded
+		// parallelism.
```

Comment:
given typechecking is IO bound, surely this function is faster? Can't we just do unbounded parallelism to simplify?